### PR TITLE
Implement JsValue to str.

### PR DESCRIPTION
**Pull request by mattcollier**
Happy New Year!

This may be a terrible idea to begin with, but if someone can help me figure out why it doesn't work, that would be awesome!

https://doc.rust-lang.org/std/ffi/struct.CStr.html#implementations

It's returning some bytes in the neighborhood of the right memory location! :)

I suspect the correct solution, assuming there is one, is a much simplified version likely using different NAPI APIs that I'm not familiar with.

### CODE REVIEW

**Comment by mattcollier** [comment]
Alright, after adding the second commit that uses `Box::into_raw` it's working as expected.  So, the question is...  Is this a terrible idea?  Is it *actually* providing a benefit?  

I'll be doing some benchmarks in my own code that will help me determine if a performance benefit is to be had.

**Comment by sehz** [comment]
Happy New Year too!  Good to start new year with holiday hacking.

Basic issue is with that [napi_get_value_string_utf8](https://nodejs.org/api/n-api.html#n_api_napi_get_value_string_utf8) copies strings into buffer.

Which means this  code is similar to `impl JSValue<'_> for String`  except instead of mapping to String, it maps to cstr and returning ref:
```
impl JSValue<'_> for String {
    fn convert_to_rust(env: &JsEnv, js_value: napi_value) -> Result<Self, NjError> {
        env.assert_type(js_value, crate::sys::napi_valuetype_napi_string)?;

        use crate::sys::napi_get_value_string_utf8;

        let mut string_size: size_t = 0;

        napi_call_result!(napi_get_value_string_utf8(
            env.inner(),
            js_value,
            ptr::null_mut(),
            0,
            &mut string_size
        ))?;

        string_size += 1;

        let chars_vec: Vec<u8> = vec![0; string_size];
        let mut chars: Box<[u8]> = chars_vec.into_boxed_slice();
        let mut read_size: size_t = 0;

        napi_call_result!(napi_get_value_string_utf8(
            env.inner(),
            js_value,
            chars.as_mut_ptr() as *mut ::std::os::raw::c_char,
            string_size,
            &mut read_size
        ))?;

        let my_chars: Vec<u8> = chars[0..read_size].into();

        String::from_utf8(my_chars).map_err(|err| err.into())
    }
}
```

**Comment by mattcollier** [comment]
@sehz I don't fully understand your proposal to use `Buffer`, but for ergonomics I would like to have `&str` if possible.

I've looked into a couple alternatives.  I think the main thing I'm needing/wanting here is a stack allocated string as described here:

https://doc.rust-lang.org/std/str/fn.from_utf8.html

I also found this idea which might be of use: https://swc.rs/rustdoc/nom/lib/std/str/fn.from_boxed_utf8_unchecked.html

Do any of these ideas seem promising?  I don't have working prototypes for either, but if one of them looks good, I could try harder.

My use case is that these `&str` are being placed into stucts in higher level code.  I'm trying to do this to avoid another copy operation to `String`.  If `&str` is not returned at this level, I run into intractable lifetime issues in higher level code.

**Comment by mattcollier** [comment]
Ok, I found this which indicates that a `buffer` is zero copy.  I'm going to see where I can get with that:

https://github.com/nodejs/node-addon-api/issues/405#issuecomment-575590460

**Comment by mattcollier** [comment]
@sehz I just pushed a commit with a new implementation, what do you think?

File: nj-core/src/convert.rs [hunk 1]
```diff
@@ -229,6 +230,27 @@ impl JSValue<'_> for String {
     }
 }
 
+impl JSValue<'_> for &str {
```
Comment (sehz) [inline nj-core/src/convert.rs:233]: Make it fully respect lifetime:
```
impl <'a> JsValue<'a> for &'a str {
```

File: nj-core/src/error.rs [hunk 2]
```diff
@@ -15,6 +16,7 @@ pub enum NjError {
     InvalidType(String, String),
     NoPlainConstructor,
     Utf8Error(FromUtf8Error),
+    Utf8Error2(Utf8Error),
```
Comment (sehz) [inline nj-core/src/error.rs:19]: still this need?

**Review by sehz — CHANGES_REQUESTED** [review]
Looking good. Minor changes requested.
Can you also bump up crate version?

**Comment by mattcollier** [comment]
Adding the lifetime in the latest commit breaks this bit of code I have and I'm not sure if/how it can be fixed.  Adding 'a lifetimes to the parameters does not help:

```
fn get_value_ref<'a>(o: &JsObject, key: &str) -> Option<&'a str> {
    let s = o.get_property(key).unwrap().unwrap();
    match s.get_property("value").unwrap() {
        Some(v) => match "value" {
            "datatype" => {
                let d = v.get_property("value").unwrap().unwrap();
                let result = d.as_value::<&str>();
                if let Err(_e) = result {
                    return None;
                }

                Some(result.unwrap())
            }
            _ => {
                let result = v.as_value::<&str>();
                if let Err(_e) = result {
                    return None;
                }

                Some(result.unwrap())
            }
        },
        None => None,
    }
}
```

**Comment by sehz** [comment]
You can write as this:
```
fn get_value_ref<'a>(o: &'a JsObject, key: &'a str) -> Option<&'a str>
```

**Comment by mattcollier** [comment]
Also, I was hoping to see a performance improvement from using this technique over using a `String`, especially when the size of the String/Buffer increases, however the performance is exactly the same.  Any thoughts on why that would be?

**Comment by mattcollier** [comment]
Well with or without the added lifetime as you specified, I get this: 

This code *does* work with the anonymous lifetime used in the earlier commit in this module.

Adding lifetime `d.as_value::<&'a str>` produces `d does not life long enough`.

```
   Compiling rust-bindgen-canonize v0.1.0 (/home/matt/dev/rust-node-bindgen-canonize)
error[E0515]: cannot return value referencing local variable `d`
  --> src/lib.rs:93:17
   |
88 |                 let result = d.as_value::<&str>();
   |                              - `d` is borrowed here
...
93 |                 Some(result.unwrap())
   |                 ^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function

error[E0515]: cannot return value referencing local variable `v`
   --> src/lib.rs:101:17
    |
96  |                 let result = v.as_value::<&str>();
    |                              - `v` is borrowed here
...
101 |                 Some(result.unwrap())
    |                 ^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function
```

**Comment by sehz** [comment]
Rust codes are very efficient so won't see perf difference unless buffers are very large.   Also Node/V8 could have slow path, so it's hard to tell unless profiling is done.

Issue with code is that `JSObject.get_property` returns owned object not reference which means you can't return string references.

**Comment by sehz** [comment]
Recommend to use our discord channel for better collaboration

**Comment by mattcollier** [comment]
> Recommend to use our discord channel for better collaboration

Sounds great, how do I get there?

**Review by sehz — APPROVED** [review]

**Pull request merged**